src/math_tools/curl.hpp

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include "ddc_alias_inline_functions.hpp"
+#include "ddc_aliases.hpp"
+#include "gradient.hpp"
+#include "metric_tensor_evaluator.hpp"
+#include "static_tensors.hpp"
+#include "vector_field.hpp"
+
+/**
+ * @brief A class which implements a curl operator
+ * The implemented equation is:
+ * @f$ \nabla \cross \mathbf{F} @f$
+ * @f$ \nabla \cross \mathbf{F} = {\cal J}_{\rm x}^{-1}\epsilon^{klm}\partial_{x^l} F_m @f$
+ * with @f$ {\cal J}_{\rm x} @f$ the jacobian of the system,
+ * @f$ F_m @f$ the covariant components of F and @f$\epsilon^{klm} @f$ the Levi-Civita symbol.
+ * @tparam Mapping3D A type representing a mapping in 3 dimensions.
+ * @tparam MappingCoord The type of the coordinate that will be used to evaluate the
+ *                  mapping. This coordinate is used to calculate the determinant of
+ *                  the Jacobian and the metric tensor. It is almost always equal to
+ *                  the argument type of the 3D mapping but axi-symmetry may make it
+ *                  useful to evaluate the mapping on a coordinate with a reduced
+ *                  dimensionality.
+ */
+template <class Mapping3D, class MappingCoord = typename Mapping3D::CoordArg>
+class Curl
+{
+    static_assert(is_mapping_v<Mapping3D>);
+
+    using BasisSpatial = ddc::to_type_seq_t<typename Mapping3D::CoordArg>;
+    using CovBasisSpatial = get_covariant_dims_t<BasisSpatial>;
+
+private:
+    Mapping3D m_mapping;
+    MetricTensorEvaluator<Mapping3D, MappingCoord> m_metric_tensor;
+
+public:
+    /**
+     * @brief Build a curl operator.
+     * @param mapping The mapping describing the system of coordinates on which the
+     *              expression is calculated.
+     */
+    explicit Curl(Mapping3D const& mapping) : m_mapping(mapping), m_metric_tensor(mapping) {}
+
+    /**
+     * @brief Compute a curl at a given coordinate, from
+     * the partial derivatives of the vector field f
+     *
+     * @param[in] partial_derivatives_f A tensor containing the partial derivatives
+     * of the vector field f expressed at the given coordinate.
+     * @param[in] coord The coordinate where the calculation is carried out.
+     */
+    template <
+            class TensorType,
+            class = std::enable_if_t<is_tensor_type_v<TensorType> && TensorType::rank() == 2>>
+    KOKKOS_INLINE_FUNCTION auto operator()(
+            TensorType const& partial_derivatives_f,
+            MappingCoord const& coord) const
+    {
+        double J = m_mapping.jacobian(coord);
+        LeviCivitaTensor<double, BasisSpatial> eps(J);
+        return tensor_mul(index<'k', 'l', 'm'>(eps), index<'l', 'm'>(partial_derivatives_f));
```

Comment:
According to the [Coding Standards](https://gyselax.github.io/gyselalibxx/docs/standards/CODING_STANDARD.html)
> The auto keyword should be avoided where possible. It is permitted in one of the following cases:
> - to store objects that can not be typed otherwise (e.g. lambdas)
> - to store the result of an expression specifically specifying the type of the generated value

Reply:
I just copy pasted from https://github.com/gyselax/gyselalibxx/blob/9a4e83ae67b2b99c8e1582e48cb88e486aa2d240/src/math_tools/lie_poisson_bracket.hpp#L65-#L73

I thought there is a specific reason to use auto. If not, I can hardcode the type

Reply:
Oops, probably I used auto in my first implementation and forgot to go back. This should have been picked up in review :|

Reply:
If you want your class to work on the cartesian basis you may need to do the same

Reply:
That makes sense. In my tests, I do not use the identity tensor

Reply:
Actually it is not even only in the tests. There are multiple places in Gysela where we have $\\{F, \hat{e}_i\\}$ where $\hat{e}_i$ are the unit vectors so I needed this use case for Lie-Poisson